pandas/core/internals/construction.py

```diff
@@ -460,13 +462,23 @@ def dict_to_mgr(
         keys = list(data.keys())
         columns = Index(keys) if keys else default_index(0)
         arrays = [com.maybe_iterable_to_list(data[k]) for k in keys]
-        arrays = [arr if not isinstance(arr, Index) else arr._data for arr in arrays]
+        if not using_copy_on_write():
+            arrays = [
+                arr if not isinstance(arr, Index) else arr._data for arr in arrays
+            ]
 
     if copy:
         if typ == "block":
             # We only need to copy arrays that will not get consolidated, i.e.
             #  only EA arrays
-            arrays = [x.copy() if isinstance(x, ExtensionArray) else x for x in arrays]
+            arrays = [
+                x.copy()
+                if isinstance(x, ExtensionArray)
+                else x.copy(deep=True)
+                if isinstance(x, Index)
```

Comment:
Yeah, this is currently a "bug" for when passing a dict of Series with EA dtype. Because we are passing a dict, it should result in a copy, but it doesn't when it's an EA dtype and does when default numpy dtype:

```
In [17]: ser = pd.Series([1, 2, 3], dtype="Int64")

In [18]: df = pd.DataFrame({"col": ser})

In [19]: df.iloc[0, 0] = 10

In [20]: ser
Out[20]: 
0    10
1     2
2     3
dtype: Int64

In [21]: ser = pd.Series([1, 2, 3], dtype="int64")

In [22]: df = pd.DataFrame({"col": ser})

In [23]: df.iloc[0, 0] = 10

In [24]: ser
Out[24]: 
0    1
1    2
2    3
dtype: int64
```

And also when creating it from an array instead of Series, it is copied correctly:

```
In [25]: arr = pd.array([1, 2, 3], dtype="Int64")

In [26]: df = pd.DataFrame({"col": arr})

In [27]: df.iloc[0, 0] = 10

In [28]: arr
Out[28]: 
<IntegerArray>
[1, 2, 3]
Length: 3, dtype: Int64
```

Reply:
Maybe we should change the check for ExtensionArray to check for the dtype being ExtensionDtype? (or move the copy deeper down into `arrays_to_mgr` after we know the data have been sanitized / unboxed.

Reply:
Would like to do this independently of the CoW change, can do as a follow up or pre-cursor. What do you prefer?

Reply:
Either way is fine for me